src/main/scala/com/mwz/sonar/scala/scoverage/ScoverageSensor.scala

```diff
@@ -0,0 +1,160 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package com.mwz.sonar.scala.scoverage
+
+import com.mwz.sonar.scala.Scala
+import com.mwz.sonar.scala.util.JavaOptionals._
+import org.sonar.api.batch.fs.{FileSystem, InputComponent, InputFile}
+import org.sonar.api.batch.sensor.{Sensor, SensorContext, SensorDescriptor}
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Loggers
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+/**
+ *  Main sensor for importing Scoverage reports to SonarQube.
+ *
+ *  @author BalmungSan
+ */
+final class ScoverageSensor extends ScoverageSensorInternal with ScoverageReportParser
+
+/** Implementation of the sensor */
+private[scoverage] class ScoverageSensorInternal extends Sensor {
+  //cake pattern to mock the scoverage report parser in tests
+  scoverageReportParser: ScoverageReportParserAPI =>
+
+  private[this] val logger = Loggers.get(classOf[ScoverageSensorInternal])
+
+  /** Populates the [[SensorDescriptor]] of this sensor. */
+  override def describe(descriptor: SensorDescriptor): Unit = {
+    descriptor
+      .onlyOnLanguage(Scala.Key)
+      .onlyOnFileType(InputFile.Type.MAIN)
+      .name(ScoverageSensorInternal.SensorName)
+  }
+
+  /** Saves in SonarQube the scoverage information of a module */
+  override def execute(context: SensorContext): Unit = {
+    logger.info("[scoverage] Initializing the scoverage sensor")
+    val reportFilename = getScoverageReportFilename(context.config())
+    Try(scoverageReportParser.parse(reportFilename)) match {
+      case Success(moduleCoverage) => {
+        logger.info(s"[scoverage] Successfully loaded the scoverage report file: '${reportFilename}'")
+
+        logger.debug("[scoverage] Saving the overall scoverage information of the module")
+        saveComponentScoverage(context, context.module(), moduleCoverage.moduleScoverage)
+
+        //save the coverage information of each file of the module
+        for (file <- getModuleSourceFiles(context.fileSystem())) {
+          //toString returns the project relative path of the file
+          val filename = file.toString
+          logger.debug(s"[scoverage] Saving the scoverage information of the file: '${filename}'")
+          moduleCoverage.filesCoverage.get(filename) match {
+            case Some(fileCoverage) => {
+              //save the file overall scoverage information
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Space after the start of the comment [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.sonar-scala.com/coding_rules#rule_key=Scalastyle%3Ascalastyle_SpaceAfterCommentStartChecker)